src/helpers/UniV3PoolBalanceCalculator.ts

```diff
@@ -41,46 +41,83 @@ export class UniV3PoolBalanceCalculator {
       const positionAsset0 = positionData[2];
       const positionAsset1 = positionData[3];
       const positionFee = positionData[4];
-      const positionliquidity = new BigNumber(positionData[7]._hex);
+      const positionLiquidity = new BigNumber(positionData[7]._hex);
+      if (positionLiquidity.isEqualTo(0)) {
+        continue;
+      }
+
+      const tickLower = positionData[5];
+      const tickUpper = positionData[6];
+
       const poolAddress = await this.uniV3PoolProvider.getPoolAddress(
         positionAsset0,
         positionAsset1,
         positionFee,
       );
 
-      // Get the total liquidity for the pool
-      const poolLiquidty = new BigNumber(
-        (
-          await this.uniV3PoolProvider.getTotalLiquidityForPool(poolAddress)
-        )._hex,
+      const poolSlot0 = await this.uniV3PoolProvider.getSlot0(poolAddress);
+      const sqrtPriceX96 = new BigNumber(poolSlot0[0]._hex);
+      const Q96 = new BigNumber(2).exponentiatedBy(96);
+
+      // Calculate the current tick
+      // Formula: currentTick = log((sqrtPrice / Q96) ** 2) / log(1.0001)
+      const currentTick = Math.floor(
+        Math.log((sqrtPriceX96.toNumber() / Q96.toNumber()) ** 2) /
+          Math.log(1.0001),
       );
 
-      // If the pool liquidity is 0, then we can't calculate the holdings so we skip but log it
-      if (poolLiquidty.isEqualTo(0)) {
-        Sentry.captureMessage(
-          `Pool liquidity is 0 for pool ${poolAddress} with position ${positions[i]}`,
+      const sqrtRatioLower = Math.sqrt(1.0001 ** tickLower);
+      const sqrtRatioUpper = Math.sqrt(1.0001 ** tickUpper);
+      const sqrtPrice = sqrtPriceX96.dividedBy(Q96);
+
+      let amount0 = new BigNumber(0);
+      let amount1 = new BigNumber(0);
+
+      // Formula for all cases comes from
+      // https://atiselsts.github.io/pdfs/uniswap-v3-liquidity-math.pdf
```

Comment:
As discussed, would be great to add some comments here about the way the calculation works for the different scenarios

Reply:
Left a comment above the function explaining all the steps and formulas.
Do you think they are sufficient?